enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,327 @@
+
+#### Scaling the StatefulSet Up/Down
+
+Once the secret is generated, users can scale the StatefulSet with no change to the usual process.
+
+#### Using TLS to Guarantee Individual Identity of StatefulSet Pods, or StatefulSets
+
+To ensure that a compromised StatefulSet pod can’t impersonate other pods from the
+same StatefulSet,
+or any other pods attached ot the same headless service
+(if this goal is achievable for the application,
+i.e. if the pods treat the intra-StatefulSet communication as untrusted
+and can plausibly protect against a rogue StatefulSet member):
+
+- Use something else, not the certificates generated by the Service CA.
+- **Don’t accept the Service CA as a valid root of trust for connections intended
+  for individual StatefulSet pods**.
+
+#### Certificate Expiration, Manual Secret Regeneration, CA Rotations
+
+No change to existing Service CA use.
+
+### Implementation Details/Notes/Constraints
+
+#### Not Generating Certificates for `127.0.0.1`, `localhost`, or NodeIPs
+
+Giving out certificates from widely-accepted CAs (like the service CA)
+with these names makes no sense because they don’t specify any service identity,
+e.g. any compromised serviceA would get a `localhost` or nodeIP certificate
+that is acceptable for connections intended to connect to serviceB.
+
+(Reportedly some OSes do actually query `localhost` on DNS on the net:
+https://archive.cabforum.org/pipermail/public/2015-June/005673.html )
+
+If a client wants to connect to _only_ a “true localhost” service,
+a service CA-signed certificate does not provide that property.
+The best way to do that is to do it directly inside the process,
+without using TLS (and the associated cryptography overhead) at all.
+Second best is to rely on known properties of the environment,
+i.e. to assume that 127.0.0.1 is localhost-only
+and to use a non-TLS connection on that address.
+
+If a _pod_ had to use TLS for localhost for some reason, the closest to a right way to do that
+(relying on TLS to verify that only loopback connections happen) is to:
+- Create a temporary CA (specific to the individual pod for)
+- Generate, and have that CA sign a localhost certificate (specific to the individual pod)
+- Set up the CA certificate as a trusted root
+- _Irreversibly erase_ the CA private key
+- Set up the pod with the generated localhost certificate+private key.
+- (This doesn’t benefit from service-ca involvement at all.)
+
+Similarly single-purpose CA could be set up for NodeIP services
+(if it makes sense to connect to node IP addresses directly at all)
+so that the client connects to the desired NodeIP service
+instead of to any NodeIP service.
+
+### Risks and Mitigations
+
+The only mitigations for the risks below are not to use this feature,
+(and prevent untrusted users from using it — using an admission controller?)
+so that the wildcard certificate is not created at all,
+or perhaps $somehow ensure that any users/code with access to the wildcard certificate
```

Comment:
Maybe just remove this line and the next? Cert security is out of scope for this proposal.

Reply:
Dropped — the $somehow is honestly a cop-out, and it’s not possible without the admission controller to restrict who can create wildcard certificates anyway.